### PR TITLE
chore(ci): Add scripts to publish hotfix changes

### PR DESCRIPTION
**Pull request by ioanabrooks**
#### Description of changes

Step to publish a hotfix against latest major version:
* create hotfix branch from latest `Version Packages` commit hash
* make changes, push 'hotfix-pr' ( PR contains fix and bumped versions )
  *  run `yarn version:hotfix` locally to create a changeset and consume it (bump versions)
* on merge to hotfix, `publish-hotfix` will kick off and publish to npm the versions specified  in the package.json, run canaries, then pubish latest
* 'hotfix-pr' merge commit will be cherry-picked into main - when merging to main, it will trigger publish latest again, but since the versions were already publish it won't do anything.

* repeat process for previous major versions if needed.

 PR on fork: https://github.com/ioanabrooks/amplify-ui/pull/19

#### Checklist

- [ ] Have read the [Pull Request Guidelines](https://github.com/aws-amplify/amplify-ui/blob/main/CONTRIBUTING.md)
- [ ] PR description included
- [ ] Relevant documentation is changed or added (and PR referenced)
- [ ] `yarn test` passes and tests are updated/added
- [ ] No side effects or [`sideEffects`](https://github.com/aws-amplify/amplify-ui/blob/main/packages/react/CONTRIBUTING.md#code-standards) field updated

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: package.json
Comment (wlee221) [inline package.json:38]: Don't think `yarn changeset` is needed unless I'm missing something:

```suggestion
    "version:hotfix": "yarn bumpVersions && yarn build",
```

File: .github/workflows/publish-latest.yml
Comment (wlee221) [inline .github/workflows/publish-latest.yml:87]: I think we discussed having hotfix workflow (1) publish to some `@canary` tag, do (2) build test against `@canary`, and (3) publish to `latest`.

Is (1) something we're still interested in doing here?

File: .github/workflows/publish-latest.yml
Comment (wlee221) [inline .github/workflows/publish-latest.yml:62]: Thanks for the v2 bump too 😈

File: package.json
Comment (ioanabrooks) [reply package.json:38]: Running changeset is needed bc otherwise `bumpVersions` doesn't have any changeset to consume. This script is meant to be run locally before pushing the hotfix PR.

File: .github/workflows/publish-latest.yml
Comment (ioanabrooks) [reply .github/workflows/publish-latest.yml:87]: I thought this runs build tests before publishing, am I under the wrong assumption?

File: package.json
Comment (wlee221) [reply package.json:38]: Ah, gotcha, makes sense.

File: .github/workflows/publish-latest.yml
Comment (wlee221) [reply .github/workflows/publish-latest.yml:87]: Right, but it would run against `@latest`, which necessarily isn't what hotfix will publish with if I understand correctly.

Comment (ioanabrooks) [reply .github/workflows/publish-latest.yml:87]: Oh I see the `changeCanaryPackageVersion` script. So lines 84-87 do nothing, I'll remove it. As for publishing twice I'm not convinced it's the right move in a high sev situation. If the changes break canaries we would know within 15min (and we would be monitoring them anyway post hotfix release). Canaries are potentially broken in the first place, since this would be a high sev situation.

Comment (wlee221) [reply .github/workflows/publish-latest.yml:87]: Don't feel strongly on this, but wouldn't hotfix commit fix canaries if they were broken? So publishing to `@canary` would at least verify the hotfix prior to publishing to @latest. 

However, I see the concern of additional complexity (+20 minutes). @reesscot, what do you think?

Comment (wlee221) [reply .github/workflows/publish-latest.yml:87]: Technically, we can make build-test faster by parallelizing it like we do in `build-and-runtime-test.yml`.

Comment (reesscot) [reply .github/workflows/publish-latest.yml:87]: It looks like the canaries only take about 5minutes to run. I think we do want to test against a tagged version of the hotfix, otherwise we may end up _not_ fixing the canaries and have to release a hotfix again (which would take even more time)? In a high severity situation we should have already mitigated by moving the `latest` tag to the previous, so I think we should be able to take the time to run our canary tests before fully deploying the hotfix to `latest`.

Comment (ioanabrooks) [reply .github/workflows/publish-latest.yml:87]: Ah gotcha, thought we were talking about the new build system tests. What would trigger the publish to `latest` then, after canaries run? would we have to do it manually?

Comment (wlee221) [reply .github/workflows/publish-latest.yml:87]: I think it can be all in one workflow that runs on pushes to `main` with the following jobs:

1. check-changesets
2. publish to `@hotfix`
3. run canary tests against `@hotfix` (parallelized, ~5min)
4. publish to `@latest`

This means that we'll probably have to write a new workflow, but I'm personally okay with that since this is a new use case.

Comment (ioanabrooks) [reply .github/workflows/publish-latest.yml]: Done! (small correction, workflow should run on pushes to `hotfix`)

File: .github/workflows/publish-latest.yml
Comment (thaddmt) [inline .github/workflows/publish-latest.yml:8]: Smallest nit in the world: but if we're updating this list of packages this is technically missing the categorical packages right?

File: .github/workflows/publish-hotfix.yml
Comment (thaddmt) [inline .github/workflows/publish-hotfix.yml:11]: Same nit here: technically missing categorical packages here

**Review by thaddmt — APPROVED** [review]

File: .github/workflows/publish-hotfix.yml
Comment (wlee221) [inline .github/workflows/publish-hotfix.yml:44]: Nit: you can use `reusable-tagged-publish`, e.g. https://github.com/aws-amplify/amplify-ui/blob/bc5533c145da3a048533813ad893a1b0d96ea341/.github/workflows/publish-next.yml#L152-L158

File: .github/workflows/publish-hotfix.yml
Comment (wlee221) [inline .github/workflows/publish-hotfix.yml:37]:
```suggestion
      #         (1) A hot fix PR was just merged, or
```

File: .github/workflows/publish-hotfix.yml
Comment (wlee221) [inline .github/workflows/publish-hotfix.yml:38]: Wonder if case (2) is really relevant here because maintainers are instructed to add and consume changesets in one PR. That said, maybe it's just nice to leave it here if we do make a mistake?

File: .github/workflows/publish-hotfix.yml
Comment (wlee221) [inline .github/workflows/publish-hotfix.yml:154]: Do we think metric jobs are necessary? This will coincide with normal `PublishLatestFailure` metric, and wondering if we really need it if we'll have bigger tickets open for hotfixing anyway.

No strong thoughts however, can be convinced!

File: .github/workflows/publish-latest.yml
Comment (wlee221) [reply .github/workflows/publish-latest.yml:8]: Think this comment is pretty outdated now, maybe we just say

> If successful, we will publish all updated UI packages to NPM.

**Review by wlee221 — COMMENTED** [review]
No blockers, thanks for your change!

File: .github/workflows/publish-hotfix.yml
Comment (ioanabrooks) [reply .github/workflows/publish-hotfix.yml:125]: I think it's nice to have, this way we're notified of failures when publishing latest as opposed to monitoring the job, it's unlikely we are doing a release at the same time anyway, and we are publishing latest at the end of the day.

File: .github/workflows/publish-hotfix.yml
Comment (ioanabrooks) [reply .github/workflows/publish-hotfix.yml:33]: Yeah, was thinking it would be nice to have in case of mistakes

File: .github/workflows/publish-hotfix.yml
Comment (ioanabrooks) [reply .github/workflows/publish-hotfix.yml:39]: Nice! updated only for hotfix, noticed the reusable script doesn't publish latest tag :)

**Review by wlee221 — DISMISSED** [review]

File: package.json
Comment (zchenwei) [inline package.json:38]: what is the purpose of doing a full `yarn build` here? Thinks only `yarn angular build` is required?

File: .github/workflows/publish-hotfix.yml
Comment (zchenwei) [inline .github/workflows/publish-hotfix.yml:33]: @wlee221 This seems dangerous if unrelated commits sneak into the branch. Can we make `hotfix` a protected branch and require approvals before a pr is able to merge in?

**Review by zchenwei — COMMENTED** [review]
Left a question otherwise it looks good to me.

File: .github/workflows/publish-hotfix.yml
Comment (wlee221) [reply .github/workflows/publish-hotfix.yml:33]: For sure, let's make that a requirement. Maybe we can 2pr sometime before we merge this?

**Review by reesscot — APPROVED** [review]

File: .github/workflows/publish-hotfix.yml
Comment (ioanabrooks) [reply .github/workflows/publish-hotfix.yml:33]: The `hotfix` branch is disposable though remember, it's always created from the latest `Version Packages` PR and deleted at the end, I don't believe there are any dangers. I can remove the (2) comment if it's making it confusing

Comment (zchenwei) [reply .github/workflows/publish-hotfix.yml]: Makes sense

**Review by zchenwei — APPROVED** [review]

**Review by wlee221 — APPROVED** [review]

**Pull request merged**
